### PR TITLE
Improves proxy support

### PR DESCRIPTION
**Pull request by adomasven**
Commit message contains relevant information. For more background:

- Bits of code were lifted from #578, where appropriate, although I had to rework most of the proxy handling protocol to support passing around proxies as JSON objects.
- As I was preparing proxy support for connectors, I looked at improving the functionality overall. Implementing #721 for connectors gave partial support to Z4FX essentially for free.

Other:

- #34 can probably be closed as it will no longer be relevant after ongoing proxy changes
- I need more background on #556, but the changes here should simplify fixing that issue as we now have more dedicated functions to detect and handle proxied urls.

### CODE REVIEW

File: chrome/content/zotero/xpcom/connector/connector.js [hunk 1]
```diff
@@ -157,14 +157,11 @@ Zotero.Connector = new function() {
...
 		var method = options.method;
-		var sendRequest = (data === null || data === undefined)
-			? Zotero.HTTP.doGet.bind(Zotero.HTTP)
-			: Zotero.HTTP.doPost.bind(Zotero.HTTP);
 		var headers = Object.assign({
 				"Content-Type":"application/json",
 				"X-Zotero-Version":Zotero.version,
 				"X-Zotero-Connector-API-Version":CONNECTOR_API_VERSION
-			}, options.headers);
+			}, options.httpHeaders || {});
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/connector/connector.js]: Should be `options.headers` after ee798dde1684a4c98b7205dad590d83b5865e868

File: test/tests/server_connectorTest.js [hunk 2]
```diff
@@ -57,13 +57,41 @@ describe("Connector Server", function () {
...
+						"Content-Type": "application/json"
+					},
+					body: JSON.stringify({
+						uri: "https://www-example-com.proxy.example.com/article",
+						html: "<head><title>Owl</title></head><body><p>🦉</p></body>"
+					})
+				}
+			);
+			
+			Zotero.debug(response.response, 1);
```
Comment (dstillman) [inline test/tests/server_connectorTest.js]: Debug line

File: test/tests/translateTest.js [hunk 3]
```diff
@@ -717,9 +712,17 @@ describe("Zotero.Translate", function() {
...
-				}));
+				})).then(function(items) {
+					try {
+						assert.equal(items[0].getField('url'), 'https://www.example.com/');
+						itemID = items[0].id;
+						itemDeferred.resolve();
+					} catch (e) {
+						itemDeferred.reject(e);
+					}
+				});
```
Comment (dstillman) [inline test/tests/translateTest.js]: Can just yield on `saveItems`

**Comment by dstillman** [comment]
Startup error in Z4Fx:

```
zotero(4)(+0000001): SELECT * FROM proxies

zotero(1)(+0000002): NS_ERROR_NOT_AVAILABLE Exception: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [mozIStorageRow.getResultByName]

    Zotero.DBConnection.prototype.queryAsync</handler.get@chrome://zotero/content/xpcom/db.js:669:14
    Zotero.Proxy.prototype._loadFromRow@chrome://zotero/content/xpcom/proxy.js:599:25
    Zotero.Proxy@chrome://zotero/content/xpcom/proxy.js:585:2
    Zotero.Proxies</this.newProxyFromRow<@chrome://zotero/content/xpcom/proxy.js:78:15
    tryCatcher@resource://zotero/bluebird.js:5147:16
    [16]</module.exports/PromiseSpawn.prototype._promiseFulfilled@resource://zotero/bluebird.js:2090:18
    [16]</module.exports/Promise.coroutine/<@resource://zotero/bluebird.js:2183:9
    Zotero.Proxies</this.init</Zotero.Proxies.proxies<@chrome://zotero/content/xpcom/proxy.js:49:21
    Zotero.Proxies</this.init<@chrome://zotero/content/xpcom/proxy.js:49:5
    tryCatcher@resource://zotero/bluebird.js:5147:16
    [16]</module.exports/PromiseSpawn.prototype._promiseFulfilled@resource://zotero/bluebird.js:2090:18
    [22]</module.exports/Promise.prototype._settlePromise@resource://zotero/bluebird.js:3246:17
    [22]</module.exports/Promise.prototype._settlePromise0@resource://zotero/bluebird.js:3286:5
    [22]</module.exports/Promise.prototype._settlePromises@resource://zotero/bluebird.js:3365:13
    [2]</Async.prototype._drainQueue@resource://zotero/bluebird.js:257:13
    [2]</Async.prototype._drainQueues@resource://zotero/bluebird.js:267:5
    [2]</Async/this.drainQueues@resource://zotero/bluebird.js:147:9
    self.setTimeout/<.notify@resource://zotero/bluebird.js:54:5

zotero(1)(+0000000): DB column 'dotsToHyphens' not found
```

**Comment by adomasven** [comment]
Any reason why [this](https://github.com/adomasven/zotero/blob/e9189291088f9fc16a24f688be63befbb62993cf/chrome/content/zotero/xpcom/proxy.js#L597-L597) wouldn't catch it? Overloaded access not captured by try/catch?

**Comment by dstillman** [comment]
It's from this: https://github.com/adomasven/zotero/blob/e9189291088f9fc16a24f688be63befbb62993cf/chrome/content/zotero/xpcom/db.js#L674

File: chrome/content/zotero/xpcom/proxy.js [hunk 4]
```diff
@@ -595,9 +595,9 @@ Zotero.Proxy.prototype._loadFromRow = function (row) {
 	this.autoAssociate = !!row.autoAssociate;
 	this.scheme = row.scheme;
 	// Database query results will throw as this option is only present when the proxy comes along with the translator
-	try {
+	if (!row instanceof Components.interfaces.mozIStorageValueArray) {
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/proxy.js]: Pretty sure you're missing some parentheses here

Comment (adomasven) [inline chrome/content/zotero/xpcom/proxy.js]: Certainly a possibility

File: chrome/content/zotero/xpcom/db.js [hunk 5]
```diff
@@ -674,6 +674,13 @@ Zotero.DBConnection.prototype.queryAsync = Zotero.Promise.coroutine(function* (s
 						Zotero.debug(msg, 1);
 						throw new Error(msg);
 					}
+				},
+				has: function(target, name) {
+					try {
+						return !!target.getResultNyName(name);
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/db.js]: `getResultByName`

**Pull request merged**